central/metrics/aggregator/common/tracker_config.go

```diff
@@ -0,0 +1,165 @@
+	}
+	return result
+}
+
+func makeGettersMap[Finding Countable](getters []LabelGetter[Finding]) map[Label]func(Finding) string {
+	result := make(map[Label]func(Finding) string, len(getters))
+	for _, getter := range getters {
+		result[getter.Label] = getter.Getter
+	}
+	return result
+}
+
+// MakeTrackerConfig initializes a tracker configuration without any period or metrics configuration.
+// Call Reconfigure to configure the period and the metrics.
+func MakeTrackerConfig[Finding Countable](category, description string,
+	getters []LabelGetter[Finding], generator FindingGenerator[Finding],
+	gauge func(string, prometheus.Labels, int),
+) *TrackerConfig[Finding] {
+	return &TrackerConfig[Finding]{
+		category:    category,
+		description: description,
+		labelOrder:  makeLabelOrderMap(getters),
+		getters:     makeGettersMap(getters),
+		generator:   generator,
+		gauge:       gauge,
+	}
+}
+
+func (tc *TrackerConfig[Finding]) Reconfigure(ctx context.Context, registry *prometheus.Registry, filter string, cfg map[string]*storage.PrometheusMetricsConfig_Labels, period time.Duration) error {
+	mcfg, err := parseMetricLabels(cfg, tc.labelOrder)
+	if err != nil {
+		return err
+	}
+	q, err := search.ParseQuery(filter, search.MatchAllIfEmpty())
+	if err != nil {
+		return err
+	}
+	tc.SetMetricsConfiguration(q, mcfg)
+	// Force track only on reconfiguration, not on the initial tracker creation.
+	if tc.setPeriod(period) {
+		tc.track(ctx)
+	}
+	return tc.registerMetrics(registry, period)
+}
+
+// setPeriod returns true if tracker period has been reconfigured.
+func (tc *TrackerConfig[Finding]) setPeriod(period time.Duration) bool {
+	tc.metricsConfigMux.Lock()
+	defer tc.metricsConfigMux.Unlock()
+	if period > 0 {
+		if tc.ticker == nil {
+			tc.ticker = time.NewTicker(period)
+			return false
+		}
+		tc.ticker.Reset(period)
+		return true
+	}
+	if tc.ticker != nil {
+		tc.ticker.Stop()
+	}
+	return false
+}
+
+func (tc *TrackerConfig[Finding]) GetMetricsConfiguration() (*v1.Query, MetricsConfiguration) {
+	tc.metricsConfigMux.RLock()
+	defer tc.metricsConfigMux.RUnlock()
+	return tc.query, tc.metricsConfig
+}
+
+func (tc *TrackerConfig[Finding]) SetMetricsConfiguration(query *v1.Query, mcfg MetricsConfiguration) {
+	tc.metricsConfigMux.Lock()
+	defer tc.metricsConfigMux.Unlock()
+	tc.query = query
+	tc.metricsConfig = mcfg
+}
+
+func (tc *TrackerConfig[Finding]) registerMetrics(registry *prometheus.Registry, period time.Duration) error {
+	for metric, labelExpression := range tc.metricsConfig {
+		if err := metrics.RegisterCustomAggregatedMetric(string(metric), tc.description, period,
+			getMetricLabels(labelExpression, tc.labelOrder), registry); err != nil {
+			return fmt.Errorf("failed to register %s metric %q: %w", tc.category, metric, err)
+		}
+		if period > 0 {
+			log.Infof("Registered %s Prometheus metric %q", tc.category, metric)
+		}
+	}
+	if period == 0 {
+		log.Infof("Metrics collection is disabled for %s", tc.category)
+	}
```

Comment:
Let's not log at such a low level and return error instead for period == 0 - or even better, validate the period very close to the user input (when reading the configmap or env) and assume that at this low level, the period will always be positive.